src/java.base/share/classes/jdk/internal/foreign/CaptureStateUtil.java

```diff
@@ -0,0 +1,350 @@
+     *
+     * For a {@code target} method handle that takes a {@code MemorySegment} and two
+     * {@code int} parameters and returns an {@code int} value, the method returns a new
+     * method handle that is doing the equivalent of:
+     * <p>
+     * {@snippet lang = java:
+     *         private static final MemoryLayout CAPTURE_LAYOUT =
+     *                 Linker.Option.captureStateLayout();
+     *         private static final CarrierLocalArenaPools POOL =
+     *                 CarrierLocalArenaPools.create(CAPTURE_LAYOUT);
+     *
+     *         public int invoke(MethodHandle target,
+     *                           String stateName,
+     *                           int a, int b) {
+     *             try (var arena = POOL.take()) {
+     *                 final MemorySegment segment = arena.allocate(CAPTURE_LAYOUT);
+     *                 final int result = (int) handle.invoke(segment, a, b);
+     *                 if (result >= 0) {
+     *                     return result;
+     *                 }
+     *                 return -(int) CAPTURE_LAYOUT
+     *                     .varHandle(MemoryLayout.PathElement.groupElement(stateName))
+     *                         .get(segment, 0);
+     *             }
+     *         }
+     *}
+     * except it is more performant. In the above {@code stateName} is the name of the
+     * captured state (e.g. {@code errno}). The static {@code CAPTURE_LAYOUT} is shared
+     * across all target method handles adapted by this method.
+     *
+     * @param target    method handle that returns an {@code int} or a {@code long} and
+     *                  has a capturing state MemorySegment as its first parameter
+     * @param stateName the name of the capturing state member layout (i.e. "errno",
+     *                  "GetLastError", or "WSAGetLastError")
+     * @throws IllegalArgumentException if the provided {@code target}'s return type is
+     *                                  not {@code int} or {@code long}
+     * @throws IllegalArgumentException if the provided {@code target}'s first parameter
+     *                                  type is not {@linkplain MemorySegment}
+     * @throws IllegalArgumentException if the provided {@code stateName} is unknown on
+     *                                  the current platform
+     */
+    public static MethodHandle adaptSystemCall(MethodHandle target,
+                                               String stateName) {
+        // Invariants checked in the BasicKey record
+        final BasicKey basicKey = new BasicKey(target, stateName);
+
+        // ((int | long), MemorySegment)(int | long)
+        final MethodHandle basicHandle = BASIC_HANDLE_CACHE
+                // Do not use a lambda in order to allow early use in the init sequence
+                // This is equivalent to:
+                //   computeIfAbsent(basicKey, CaptureStateUtil::basicHandleFor);
+                .computeIfAbsent(basicKey, new Function<>() {
```

Comment:
I recommend a local record and capture the record instance in a member static final field. This code creates a function on every call. Also might be of interest whether we should use get + putIfAbsent or computeIfAbsent, as CHM has some bug that makes cIA slower than get for certain access patterns.